ffi/src/scan.rs

```diff
@@ -7,45 +7,54 @@ use std::sync::Arc;
 use delta_kernel::scan::state::{visit_scan_files, DvInfo, GlobalScanState};
 use delta_kernel::scan::{Scan, ScanBuilder, ScanData};
 use delta_kernel::{DeltaResult, EngineData};
+use delta_kernel_ffi_macros::handle_descriptor;
 use tracing::debug;
 use url::Url;
 
 use crate::{
-    unwrap_kernel_expression, AllocateStringFn, EnginePredicate, ExternEngine, ExternEngineHandle,
-    ExternResult, IntoExternResult, KernelBoolSlice, KernelExpressionVisitorState,
-    KernelStringSlice, NullableCvoid, SnapshotHandle, TryFromStringSlice,
+    unwrap_kernel_expression, AllocateStringFn, EnginePredicate, ExternEngine, ExternResult,
+    IntoExternResult, KernelBoolSlice, KernelExpressionVisitorState, KernelStringSlice,
+    NullableCvoid, SharedExternEngine, SharedSnapshot, TryFromStringSlice,
 };
 
-use super::handle::{ArcHandle, BoxHandle};
+use super::handle::{CloneHandle, Handle, HandleAsMut};
 
 // TODO: Do we want this type at all? Perhaps we should just _always_ pass raw *mut c_void pointers
 // that are the engine data
 /// an opaque struct that encapsulates data read by an engine. this handle can be passed back into
 /// some kernel calls to operate on the data, or can be converted into the raw data as read by the
 /// [`delta_kernel::Engine`] by calling [`get_raw_engine_data`]
-pub struct EngineDataHandle {
-    data: Box<dyn EngineData>,
-}
-impl BoxHandle for EngineDataHandle {}
+#[handle_descriptor(target=dyn EngineData, mutable=true, sized=false)]
+pub struct EngineDataHandle;
 
 /// Allow an engine to "unwrap" an [`EngineDataHandle`] into the raw pointer for the case it wants
 /// to use its own engine data format
 ///
 /// # Safety
-/// `data_handle` must be a valid pointer to a kernel allocated `EngineDataHandle`
-pub unsafe extern "C" fn get_raw_engine_data(data_handle: *mut EngineDataHandle) -> *mut c_void {
-    let boxed_data = unsafe { Box::from_raw(data_handle) };
-    Box::into_raw(boxed_data.data).cast()
+///
+/// `data_handle` must be a valid pointer to a kernel allocated `EngineDataHandle`. The Engine must
+/// ensure the handle outlives the returned pointer.
+// TODO(frj): What is the engine actually doing with this method??
+#[no_mangle]
+pub unsafe extern "C" fn get_raw_engine_data(mut data: Handle<EngineDataHandle>) -> *mut c_void {
+    let ptr = get_raw_engine_data_impl(&mut data) as *mut dyn EngineData;
+    ptr as _
+}
+
+unsafe fn get_raw_engine_data_impl(data: &mut Handle<EngineDataHandle>) -> &mut dyn EngineData {
+    let _data = unsafe { data.as_mut() };
+    todo!() // See TODO comment for EngineDataHandle
 }
 
 /// Struct to allow binding to the arrow [C Data
 /// Interface](https://arrow.apache.org/docs/format/CDataInterface.html). This includes the data and
 /// the schema.
 #[cfg(feature = "default-engine")]
-#[repr(C)]
+// FIXME: cbindgen doesn't recognize the FFI_ArrowFoo fields in this struct
```

Comment:
I fixed this here: https://github.com/delta-incubator/delta-kernel-rs/pull/203/files#diff-e4dc56c14dad42367f172aee9d547df5959ef6b1b3a4575dfc4d989c0f97ab8bR21